### PR TITLE
only allow deployments of verified users

### PR DESCRIPTION
**Pull request by Eslam-Nawara**
### Description

only allow deployments of verified users

### Related Issues

- #2444 

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

### CODE REVIEW

**Comment by Eslam-Nawara** [comment]
This pr would depend on the deployment of kyc server implemented [here](https://github.com/threefoldtech/tfgrid-kyc-verifier/pull/4) having this [requirements](https://github.com/threefoldtech/tfgrid-kyc-verifier/issues/5#issuecomment-2434853707), **it should not be merged before the deployment of the server with the required endpoint**

Merging the pr would cause that no unverified user would be able to use any zos3 to deploy any solution, **so it shouldn't be merged before the dashboard has a way to verify the twins**

Also we need to verify the twin used in github workflows so the test can run deployments

File: pkg/gridtypes/deployment.go [hunk 1]
```diff
@@ -665,3 +672,46 @@ func (o JobOperation) String() string {
 		return "unknown"
 	}
 }
+
+// getTwinVerificationState make sure the account used is verified we have the user public key in bytes(pkBytes)
+func getTwinVerificationState(twinID uint32) (status string) {
+	verificationServiceURL := "https://kyc1.gent01.dev.grid.tf/api/v1/status"
```
Comment (xmonader) [inline pkg/gridtypes/deployment.go]: this url should be retrieved from stack information

**Review by ashraffouda — DISMISSED** [review]

File: pkg/provision/engine.go [hunk 2]
```diff
@@ -1162,3 +1189,57 @@ func (e *NativeEngine) GetWorkloadStatus(id string) (gridtypes.ResultState, bool
...
+		Result struct {
+			Status string `json:"status"`
+		} `json:"result"`
+		Error string `json:"error"`
+	}
+
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return
+	}
```
Comment (muhamadazmy) [inline pkg/provision/engine.go]: You can create a json Decoder with the body and hence you will not have to read it in memory first

File: pkg/provision/engine.go [hunk 3]
```diff
@@ -1162,3 +1189,57 @@ func (e *NativeEngine) GetWorkloadStatus(id string) (gridtypes.ResultState, bool
 
 	return wl.Result.State, true, nil
 }
+
+// getTwinVerificationStatus make sure the account used is verified we have the user public key in bytes(pkBytes)
+func getTwinVerificationStatus(twinID uint32) (status string) {
+	status = "FAILED"
```
Comment (muhamadazmy) [inline pkg/provision/engine.go]: Please use a const instead. If possible values can only be success/failure then it's better to use a boolean maybe

Define a new type alias so it's not confused with actual strings

File: pkg/provision/engine.go [hunk 4]
```diff
@@ -1012,6 +1024,18 @@ func (n *NativeEngine) CreateOrUpdate(twin uint32, deployment gridtypes.Deployme
 		return fmt.Errorf("twin id mismatch (deployment: %d, message: %d)", deployment.TwinID, twin)
 	}
 
+	// make sure the account used is verified
+	check := func() error {
+		if getTwinVerificationStatus(twin) != "VERIFIED" {
```
Comment (muhamadazmy) [inline pkg/provision/engine.go]: Define constants or even better create a new type alias

File: pkg/provision/engine.go [hunk 5]
```diff
@@ -1162,3 +1189,57 @@ func (e *NativeEngine) GetWorkloadStatus(id string) (gridtypes.ResultState, bool
...
+
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return
+	}
+
+	if response.StatusCode != http.StatusOK {
+		log.Error().Msgf("failed to verify user status: %s", result.Error)
+		return
+	}
```
Comment (muhamadazmy) [inline pkg/provision/engine.go]: This check needs to be done before you actually deserialize the body, no? If status is not okay a big chance the body is not json

**Review by muhamadazmy — DISMISSED** [review]

**Review by rawdaGastan — APPROVED** [review]

**Pull request merged**